xapian-applications/omega/handler_libebook.cc

```diff
@@ -0,0 +1,133 @@
+/** @file handler_libebook.cc
+ * @brief Extract text and metadata using libe-book.
+ */
+/* Copyright (C) 2019 Bruno Baruffaldi
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include <xapian.h>
+#include <string.h>
+#include <unordered_map>
+#include <memory>
+#include <iostream>
+
+#include <librevenge-0.0/librevenge-generators/librevenge-generators.h>
+#include <librevenge-0.0/librevenge-stream/librevenge-stream.h>
+#include <libe-book/libe-book.h>
+
+using libebook::EBOOKDocument;
+using namespace std;
+using namespace librevenge;
+
+static
+char* str_copy(const char* text) {
+    int len = strlen(text);
+    char* line = new char[len];
+    strcpy(line, text);
+    return line;
+}
+
+static
+void clear_text(string& str,const char* text) {
+    int len = strlen(text);
+    for (int i = 0; i < len; ++i)
+	if (!isspace(text[i]) || (i && !isspace(text[i - 1])))
+	    str.push_back(text[i]);
+}
```

Comment:
Is it actually necessary to run every string through this function?  Unless the output we're getting has a lot of sequences of multiple spaces, any saving we get from stripping them out here is going to be cancelled out by the cost of scanning and copying these strings like this.

If we do need to do this, we should at least avoid the `strlen(text)` which unnecessary scans the string yet another time just to find its length - we can make the `for` loop's end condition check for the zero byte instead.

Reply:
Okay, I will remove strlen.
I think I can use the function clean_text only in dump. Searching time increases a lot if I didn't use this function (at least with poppler). I will try to find some ebooks to test it (with more than 500 pages).

Reply:
PDF is a very different format to these ebooks though.

Also I'm surprised to hear there's any search-time impact - the whitespace should be skipped when splitting the text into terms, and the stored sample gets whitespace cleaned by `omindex`.  I'd expect any saving you get from cleaning whitespace here to be down to the reduced traffic across the pipe between omindex and the worker process.  The one exception is `author` which is currently just stored as-is (that's really a bug, but I suspect in practice that author metadata doesn't have a lot of extra whitespace).

Reply:
I will fix handler_poppler after get this merged.